nncf/experimental/torch/nas/bootstrapNAS/search/evaluator.py

```diff
@@ -0,0 +1,178 @@
+"""
+ Copyright (c) 2022 Intel Corporation
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+      http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+from typing import Any, Tuple, NoReturn
+from typing import Callable
+from typing import Dict
+import csv
+
+from nncf.common.utils.logger import logger as nncf_logger
+from nncf.experimental.torch.nas.bootstrapNAS.elasticity.elasticity_controller import ElasticityController
+from nncf.experimental.torch.nas.bootstrapNAS.elasticity.multi_elasticity_handler import MultiElasticityHandler
+from nncf.torch.nncf_network import NNCFNetwork
+
+class BNASEvaluatorStateNames:
+    BNAS_EVALUATOR_STAGE = 'evaluator_state'
+
+def get_flops_for_active_subnet(elasticity_handler: MultiElasticityHandler) -> float:
+    flops, _ = elasticity_handler.count_flops_and_weights_for_active_subnet()
+    return flops /2000000   # MACs
+
+def get_weights_for_active_subnet(elasticity_handler: MultiElasticityHandler) -> float:
+    _, num_weights = elasticity_handler.count_flops_and_weights_for_active_subnet()
+    return num_weights
+
+class Evaluator:
+    """
+    An interface for handling measurements collected on a target device. Evaluators make use
+    of functions provided by the users to measure a particular property, e.g., accuracy, latency, etc.
+    """
+    def __init__(self, name: str, eval_func: Callable, ideal_val: float, elasticity_ctrl: ElasticityController):
+        self.name = name
+        self._eval_func = eval_func
+        self._current_value = 0
+        self._ideal_value = ideal_val
+        self._elasticity_ctrl = elasticity_ctrl
+        self._use_model_for_evaluation = True
+        self.cache = {}
+        self.input_model_value = None
+        #TODO(pablo): Here we should store some super-network signature that is associated with this evaluator
+
+    @property
+    def current_value(self):
+        return self._current_value
+
+    @current_value.setter
+    def current_value(self, val : float) -> NoReturn:
+        self._current_value = val
+
+    @property
+    def use_model_for_evaluation(self):
+        return self._use_model_for_evaluation
+
+    @use_model_for_evaluation.setter
+    def use_model_for_evaluation(self, val: bool) -> NoReturn:
+        self._use_model_for_evaluation = val
```

Comment:
Seems like it is a bad design of base evaluator. It introduces ambiguity and sometimes not expected behavior.
I'd not introduce such logic to the base class, it heavily constrains the successor classes.
Moreover, this value can be invalid for `FlopsEvaluator` - it can't evaluate using model by design. I think it shouldn't be set externally.
Let the child class define how it's going to evaluate. If it always evaluates using handler, there shouldn't be logic with evaluation using model.

Reply:
I have refactored the evaluators based on your comments. Let's have a discussion. tHanks!